src/lib/server/query_handler.cpp

```diff
@@ -0,0 +1,85 @@
+#include "query_handler.hpp"
+
+#include "expression/value_expression.hpp"
+#include "hyrise.hpp"
+#include "sql/sql_pipeline_builder.hpp"
+#include "sql/sql_translator.hpp"
+
+namespace opossum {
+
+ExecutionInformation QueryHandler::execute_pipeline(const std::string& sql,
+                                                    const SendExecutionInfo send_execution_info) {
+  // A simple query command invalidates unnamed statements
+  // See: https://postgresql.org/docs/11/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY
+  if (Hyrise::get().storage_manager.has_prepared_plan("")) Hyrise::get().storage_manager.drop_prepared_plan("");
+
+  auto execution_info = ExecutionInformation();
+  auto sql_pipeline = SQLPipelineBuilder{sql}.create_pipeline();
+  sql_pipeline.get_result_table();
+
+  const auto [pipeline_status, result_table] = sql_pipeline.get_result_table();
+  if (pipeline_status == SQLPipelineStatus::Success) {
+    execution_info.result_table = result_table;
+    execution_info.root_operator = sql_pipeline.get_physical_plans().back()->type();
+
+    if (send_execution_info == SendExecutionInfo::Yes) {
+      std::stringstream stream;
+      stream << sql_pipeline.metrics();
+      execution_info.pipeline_metrics = stream.str();
+    }
+  } else {
+    const std::string failed_statement = sql_pipeline.failed_pipeline_statement()->get_sql_string();
+    execution_info.error = "Error during pipeline execution. Failed statement: " + failed_statement;
```

Comment:
So what is the right way to handle this? Or is it just bad a name?

Reply:
Hmm, can you return SQL STATE 40001? That looks closest to what postgres is doing. Also please check for `SQLPipelineStatus::RolledBack` and change the text to "Transaction conflict, transaction was rolled back.".

Reply:
Returning SQL codes that can be used by clients (such as psql) would require to implement another error message format. In order to keep the code clear I'd stick to the string solution.

You are referring to transactions. Does this (`SQLPipelineStatus::RolledBack`) only happen within a transaction context? So far the server does not support transactions since the parser doesn't understand "begin", "commit" and "rollback".

Reply:
> So far the server does not support transactions since the parser doesn't understand "begin", "commit" and "rollback".

Every SQL Pipeline uses transactions unless they have been explicitly disabled. Otherwise, we could not perform any data modifications. If no transaction context has been supplied to the builder, one is started for each statement:

https://github.com/hyrise/hyrise/blob/90aaf912a1fdeb335417ae09b9e5a9b95d1c80b3/src/lib/sql/sql_pipeline_statement.cpp#L144

> You are referring to transactions. Does this (SQLPipelineStatus::RolledBack) only happen within a transaction context?

Yes. Running a statements without a transaction context is only legal for read-only statements; those are guaranteed to not conflict.

> Returning SQL codes that can be used by clients (such as psql) would require to implement another error message format. In order to keep the code clear I'd stick to the string solution.

Once we execute multiple modifying statements concurrently (e.g., in the Bachelor's project), we will definitely run into that issue. This is very different from an actual error as a transaction conflict is something that is unwanted but expected. As such, we should not have to resort to string parsing to distinguish conflicts from actual errors.

Can you estimate how much of a change this is? I don't want to burden you with this - if it is more than a couple of lines, we can add it as a todo. In that case, maybe you could outline what needs to happen?

Reply:
Thanks for the explanation. Shouldn't be too difficult, since this is documented. Gonna have a look at this tomorrow.